### DISCUSSION TITLE
Quest 3 detected (or reported) as Quest Pro, low resolution?

### DISCUSSION THREAD

**Discussion by paulmelis** [StereoKit/StereoKit#926] [Q&A]
Just noticed this in logcat output:

```
03-27 11:52:12.326 12297 12322 D StereoKit: Initializing StereoKit v0.3.9 Android ARM64...
03-27 11:52:12.326 12297 12322 D StereoKit: Initializing Platform
03-27 11:52:12.352 12297 12322 D StereoKit: [sk_gpu] EGL version 1.5
03-27 11:52:12.355 12297 12322 D StereoKit: [sk_gpu] Couldn't load gl function glPolygonMode
03-27 11:52:12.357 12297 12322 D StereoKit: [sk_gpu] Couldn't load gl function glGetTexImage
03-27 11:52:12.358 12297 12322 D StereoKit: [sk_gpu] Using OpenGL: OpenGL ES 3.2 V@0757.0 (GIT@08435ba2e8, I4fef8e100e, 1708703667) (Date:02/23/24)
03-27 11:52:12.358 12297 12322 D StereoKit: [sk_gpu] Device: Adreno (TM) 740
03-27 11:52:12.358 12297 12322 D StereoKit: Starting a XR mode app
...
03-27 11:52:12.415 12297 12322 D StereoKit: | ACTIVATED | XR_EXT_hand_tracking
03-27 11:52:12.415 12297 12322 D StereoKit: | ACTIVATED | XR_EXT_hand_tracking_data_source
03-27 11:52:12.415 12297 12322 D StereoKit: | ACTIVATED | XR_EXT_local_floor
03-27 11:52:12.415 12297 12322 D StereoKit: | ACTIVATED | XR_FB_color_space
03-27 11:52:12.415 12297 12322 D StereoKit: | ACTIVATED | XR_KHR_android_create_instance
03-27 11:52:12.415 12297 12322 D StereoKit: | ACTIVATED | XR_KHR_composition_layer_depth
03-27 11:52:12.415 12297 12322 D StereoKit: | ACTIVATED | XR_KHR_convert_timespec_time
03-27 11:52:12.415 12297 12322 D StereoKit: | ACTIVATED | XR_KHR_opengl_es_enable
03-27 11:52:12.415 12297 12322 D StereoKit: |___________|______________________
03-27 11:52:12.457 12297 12322 D StereoKit: Found OpenXR runtime: Oculus 63.325.0
03-27 11:52:12.458 12297 12322 D StereoKit: OpenXR requires GLES v3.0.0 - v3.2.0
03-27 11:52:12.709 12297 12322 D StereoKit: System name: Meta Quest Pro <<<<<<<<<<<<<<<<<<<<
03-27 11:52:12.709 12297 12322 D StereoKit: Platform does not support single-pass rendering
03-27 11:52:12.709 12297 12322 D StereoKit: OpenXR articulated hands ext enabled!
03-27 11:52:12.709 12297 12322 D StereoKit: Set color space to XR_COLOR_SPACE_REC709_FB
03-27 11:52:12.709 12297 12322 D StereoKit: Creating view: PrimaryStereo color:rgba32_sRGB depth:depth16 blend:Opaque
03-27 11:52:12.722 12297 12322 D StereoKit: Set view: PrimaryStereo to 1680x1760@1msaa <<<<<<<<<<<<<<<<<<<<
03-27 11:52:12.744 12297 12322 D StereoKit: OpenXR session begin.
```

Is this the OpenXR layer reporting the device as a Quest Pro? And with a low resolution of 1680x1760 (actual Quest 3: 2064x2208 per eye, Quest Pro: 1800x1920 per eye)?

I know about `SKSettings.renderScaling` being [needed](https://github.com/StereoKit/StereoKit/discussions/844) on a Quest, but I'm surprised the device basically gets misrepresented.

**Comment by mvvvv** [comment]
Yes, openxr is a bit disappointing on Quests , especially on anti-aliasing. Nick replied this:
https://github.com/StereoKit/StereoKit/discussions/844#discussioncomment-8184030

I didn't see any change with v63 on Quest 2

**Comment by paulmelis** [comment]
`Platforms/Android/AndroidManifest.xml` (as written by `dotnet new sk-multi`) contained no reference to Quest 3:

```
		<!-- Oculus -->
		<meta-data android:name="com.oculus.supportedDevices" android:value="quest|quest2|questpro"/>
```

Ammending it to include `quest3`:
```
		<meta-data android:name="com.oculus.supportedDevices" android:value="quest|quest2|questpro|quest3"/>
```

That at least makes the device identify itself as Quest 3 :)

```
03-27 15:20:18.682  7657  7683 I StereoKit: Initialization successful
03-27 15:20:18.688  7657  7683 I StereoKit: Device: Meta Quest 3
03-27 15:20:18.691  7657  7683 I StereoKit: Display resolution 1680x1760
```

No change in resolution, though.

**Reply by maluoi** [reply]
The current template does have `quest3` in the manifest! I think I slept a little on adding it though, so it's a relatively recent addition.

**Reply by paulmelis** [reply]
Ah, didn't notice that! It was a bit of an older project I was testing here

**Comment by paulmelis** [comment]
Oh, interesting, from https://communityforums.atmeta.com/t5/Talk-VR/The-Quest-3-s-standalone-render-resolution-is-only-1680x1760/td-p/1090155 

> Thanks to Norm for being the only reviewer so far I've seen to point this out. While the Q3's panels have a native resolution of 2064x2208, for standalone use things are only rendered at 1680x1760 resolution. For standalone, graphics aren't rendered at the native resolution of the panels. It's like watching a 720p video on a 1080p screen. To get full native resolution you have to use it with a PC and supersample.

**Comment by maluoi** [comment]
Quest has always used a pretty low resolution compared to the size of its actual panels. It took me a while to settle on why exactly they do this, but I think the core reason is that they've decided 4x MSAA should be standard for XR apps! I think they did some studies and found 4xMSAA at a lower resolution is the right tradeoff to make for performance/quality. I feel like they had a page or two in their docs on the subject, but all I'm finding right now is [this one](https://developer.beta.oculus.com/documentation/native/android/mobile-msaa-analysis/).

I _almost_ had MSAA working on Quest, but ran into a tough GLES bug. It was taking too long to track it down, so I unfortunately had to shift it to the backburner for a while. It's a higher priority item right now, so I will be re-visiting it soon.

**Reply by paulmelis** [reply]
It's really weird to see rendering on a Magic Leap 2, at a reported (by SK) resolution of 1440x1760, be so much nicer without any render scaling or MSAA enabled (at least not by me). Is that purely due to a different display technology, causing a "natural" form of anti-aliasing?

**Reply by maluoi** [reply]
Yeah the big reason why SK doesn't have MSAA is because most devices look great without it! On PC, OpenXR runtimes typically use a rendering surface larger than their display to compensate for distortion, so you essentially get supersampling.

HoloLens 2 looked really good too, I don't remember if they were super-sampling as well though. Those waveguide displays are pretty complicated though, so it's tough to say what was really hiding it. I know HL2 did interlacing and was kinda fuzzy to begin with, I kinda do think those waveguides are just not crisp in the same way, so it hides aliasing better?

At a minimum, they're rendering much closer to their target resolution.

**Comment by paulmelis** [comment]
Should `SKSettings.renderScaling` (still) work on Quest 3 with SK 0.3.9? I've tested with 1.5 and 2.0, but really don't see any difference. Certainly no clear reduction in aliasing. I do see a higher resolution being reported by `SK.System.displayWidth`/`Height`, just no visual improvement.

**Reply by maluoi** [reply]
Oh hey, yeah it looks like `SKSettings.renderScaling` is getting overwritten at some point during initialization! I'll look into that. In the meantime, if you use `Renderer.Scaling` right after init or something, that's still working fine.

**Reply by paulmelis** [reply]
Oh yeah, now rendering is *much* better!

**Reply by mvvvv** [reply]
now it works for me as well. ^_^ . Only the scaling but not the multisample which is always @1MSAA
